homeassistant/components/lifx/coordinator.py

```diff
@@ -188,42 +189,61 @@ def async_get_entity_id(self, platform: Platform, key: str) -> str | None:
 
     async def _async_update_data(self) -> None:
         """Fetch all device data from the api."""
-        async with self.lock:
-            if self.device.host_firmware_version is None:
-                self.device.get_hostfirmware()
-            if self.device.product is None:
-                self.device.get_version()
-            if self.device.group is None:
-                self.device.get_group()
-
-            response = await async_execute_lifx(self.device.get_color)
-
-            if self.device.product is None:
-                raise UpdateFailed(
-                    f"Failed to fetch get version from device: {self.device.ip_addr}"
+
+        for _ in range(MAX_TIMEOUTS_TO_DECLARE_UPDATE_FAILED - 1):
+            try:
+                tasks: list[Awaitable] = [async_execute_lifx(self.device.get_color)]
+
+                if self.device.host_firmware_version is None:
+                    tasks.append(async_execute_lifx(self.device.get_hostfirmware))
+                if self.device.product is None:
+                    tasks.append(async_execute_lifx(self.device.get_version))
+                if self.device.group is None:
+                    tasks.append(async_execute_lifx(self.device.get_group))
+
+                await asyncio.gather(*tasks)
+
+                while len(self.device.message) > 0:
+                    # let the loop run until all messages have replies or aiolifx times out waiting for them
+                    await asyncio.sleep(0)  # pragma: no cover
```

Comment:
I think it's still possible to write a test for this if you patch the lib with a side effect that returns returns a value the first time and nothing on subsequent calls

Reply:
> Looks good. We should cover that line though. If you get stuck I can help with the test later tonight

I'll take all the help I can get: tests are still something I'm getting my head around and I have no idea how to mock messages in the queue such that they would survive at least one sleep, etc.

Though, I suspect this test and the `cant_connect` lines we're missing in `config_flow.py` would have a similar solution.

Reply:
```

async def test_updates_wait_for_pending_messages(hass: HomeAssistant) -> None:
    """Test updates wait for pending messages."""
    already_migrated_config_entry = MockConfigEntry(
        domain=DOMAIN, data={CONF_HOST: IP_ADDRESS}, unique_id=SERIAL
    )
    already_migrated_config_entry.add_to_hass(hass)
    light_strip = _mocked_light_strip()
    entity_id = "light.my_bulb"
    calls = 0

    class MockMessageDict(dict):

        def __len__(self):
            """Return the length of the messages."""
            nonlocal calls
            calls += 1
            if calls == 1:
                return 1
            return 0
    
    _mock_message_dict = MockMessageDict()

    with patch.object(light_strip,"message",_mock_message_dict), _patch_discovery(device=light_strip), _patch_device(device=light_strip):
        await async_setup_component(hass, lifx.DOMAIN, {lifx.DOMAIN: {}})
        await hass.async_block_till_done()
        entity_registry = er.async_get(hass)
        assert entity_registry.async_get(entity_id).unique_id == SERIAL
        assert hass.states.get(entity_id).state == STATE_OFF
        assert calls == 2
```

This will likely do what you want, but there is a problem with waiting for a `dict` to be empty.  If something goes wrong with `aiolifx` and the `dict` never gets emptied or it takes more than one wait this gets expensive very quickly as we either have event loop thrashing where it keeps coming back and checking it and returning control to the loop or a problem where it never gets emptied and we effectively have an infinite loop.

Reply:
Thanks! I'll give it a whirl later and see how it goes.

Reply:
I think you could solve the loop problem by waiting on the events  with something like this



```python
if sys.version_info[:2] < (3, 11):
    from async_timeout import timeout as asyncio_timeout
else:
    from asyncio import timeout as asyncio_timeout
```

```python
                while self.device.message:
                    events = []
                    # Make a copy of the list to avoid a the dictionary changing
                    # size during iteration
                    for _response, event, _callb in list(self.device.message.values()):
                        events.append(event.wait())
                    async with asyncio_timeout(MESSAGE_TIMEOUT):
                        await asyncio.gather(*events)
```

Also I expect one of the reason the lib is unreliable is it uses `asyncio.wait_for` on `event.wait()` which had race problems that aren't fixed until python 3.12. Removing all the `wait_for`s in `aiolifx` and replacing them with `asyncio_timeout` will likely make the lib more reliable.

Reply:
Yeah, I'd love to refactor it to get rid of all the callbacks too. There is actually fork that has already done it, but it's old and thus is missing lots of the new products and messages.

Reply:
I'll apply both and see if anything explodes.

Reply:
Ok, this is _way_ slower. Updates go from ~0.1s to ~2.5 and I'm getting timeouts almost immediately.

Reply:
We need another solution as we can't use a len check on a dict as a synchronization method without the above risk as this is effectively implementing a lock that waits for all messages to be answered without having a proper waiter implementation

Reply:
I deleted my initial reply as it was user error. The replacement works, there's just a line still uncovered by the tests.

Reply:
I'm running out of time to look at this for the day, but I'll try to take a look tomorrow

Reply:
I'll keep poking at it with a blunt stick for a while to see if that helps.

Reply:
I've gone and made things complicated by adding a check for the type of message because `aiolifx` doesn't create events for the `Acknowledgement` message type. Which means our test needs to have mocked messages so that the check is covered. I'll commit the change to `coordinator.py` so you can see what I'm talking about.